datadriven/src/datadriven.rs

```diff
@@ -71,233 +78,105 @@ use std::path::{Path, PathBuf};
 /// }
 /// ```
 ///
-pub fn run_test<F>(path: &str, f: F, logger: &slog::Logger) -> Result<()>
+/// Rewrite Mode:
+///
+/// When the testdata is changed, you can run the rewrite mode directly to get the new testdata.
+/// If there are no blank line in expected value, it will be overwritten with one separator,
+/// instead of double separator, vice versa.
+///
+pub fn run_test<F>(path: &str, f: F, rewrite: bool, logger: &slog::Logger) -> Result<()>
 where
     F: FnOnce(&TestData) -> String + Copy,
 {
-    match fs::read_dir(path) {
-        Ok(read_dir) => {
-            let entries: Vec<PathBuf> = read_dir
-                .map(|res| res.map(|e| e.path()))
-                .collect::<Result<Vec<_>, std::io::Error>>()?;
-
-            for path in entries.iter() {
-                let file = fs::read_to_string(path)?;
-                run_test_internal(path, &file, f, logger)?;
-            }
-        }
-        _ => {
-            let file = fs::read_to_string(path)?;
-            run_test_internal(path, &file, f, logger)?;
+    let files = get_dirs_or_file(path)?;
+
+    for path in &files {
+        let content = fs::read_to_string(path)?;
+        if let Some(rewrite_data) = run_test_internal(path, &content, f, rewrite, logger)? {
+            let mut file = OpenOptions::new().write(true).truncate(true).open(path)?;
+            file.write_all(rewrite_data.as_bytes())?;
+            file.sync_data()?;
+        } else {
+            // test mode, skip rewriting data
         }
     }
 
     Ok(())
 }
 
-fn run_test_internal<F, P>(source_name: P, content: &str, f: F, logger: &slog::Logger) -> Result<()>
+fn run_test_internal<F, P>(
+    source_name: P,
+    content: &str,
+    f: F,
+    rewrite: bool,
+    logger: &slog::Logger,
+) -> Result<Option<String>>
 where
     F: FnOnce(&TestData) -> String + Copy,
     P: AsRef<Path>,
 {
-    let mut r = TestDataReader::new(source_name, content, logger);
+    let mut r = TestDataReader::new(source_name, content, rewrite, logger);
+
     while r.next()? {
-        run_directive(&r, f)?;
+        run_directive(&mut r, f)?;
     }
-    Ok(())
+
+    // remove redundant '\n'
+    let data = r.rewrite_buffer.map(|mut rb| {
+        if rb.ends_with("\n\n") {
+            rb.pop();
+            rb.pop();
```

Comment:
Why? I can see go implement will still keep one newline.

Reply:
There is no "\n" at the end of rewrite data, so I need to pop 2 "\n"

Reply:
the addition "\n" is the reason why I pop 2 "\n"

Reply:
I am not sure why they dont have any blank line at the end